File: python/euler/bin/p0098.py
# project euler: problem 98

from collections.abc import Iterable, Iterator
from itertools import combinations, count
from math import isqrt
from typing import TextIO

from euler.lib.util import assoc_group_dict, num_of_digits


def select_keywords(words: Iterable[str]) -> tuple[int, list[tuple[str, list[str]]]]:
    def make_key(word: str) -> str:
        return ''.join(sorted(list(word)))

    def make_kw_info(
        tpl_it: Iterator[tuple[str, str]],
    ) -> tuple[int, list[tuple[str, list[str]]]]:
        work = assoc_group_dict(tpl_it)
        del_list = []
        for k, v in work.items():
            if len(v) == 1:
                del_list.append(k)
        for k in del_list:
            del work[k]

        result = sorted(list(work.items()), key=lambda x: len(x[0]), reverse=True)
        max_digits = len(result[0][0])

        return (max_digits, result)

    return make_kw_info(map(lambda word: (make_key(word), word), words))


def make_sq_tbl(max_digits: int) -> tuple[dict[int, list[str]], dict[int, list[str]]]:
    limit = isqrt((10**max_digits) - 1)
    work = []
    for n in count(1):
        if n > limit:
            break
        sq = n * n
        work.append((num_of_digits(sq), str(sq)))

    sq_tbl = assoc_group_dict(work)
    sq_uniq_tbl = dict((k, list(filter(lambda x: len(x) == len(set(x)), v))) for k, v in sq_tbl.items())

    return (sq_tbl, sq_uniq_tbl)


def compute(fh: TextIO) -> str:
    def check_pair(w1: str, w2: str) -> int | None:
        ndigits = len(w1)
        tbl = sq_uniq_tbl if ndigits == len(set(w1)) else sq_tbl

        for sq in tbl[ndigits]:
            pair_dict = assoc_group_dict(zip(w1, sq, strict=True))
            if (
                all([len(v) == 1 for v in pair_dict.values()])
                and (w2_trans := w2.translate(str.maketrans(w1, sq))) in tbl[ndigits]
            ):
                return max(int(w1.translate(str.maketrans(w1, sq))), int(w2_trans))

        return None

    words = [s.strip('"') for s in fh.read().split(',')]
    max_digits, kw = select_keywords(words)
    sq_tbl, sq_uniq_tbl = make_sq_tbl(max_digits)

    for _, v in kw:
        for w1, w2 in combinations(v, 2):
            if (result := check_pair(w1, w2)) is not None:
                return str(result)

    raise RuntimeError('unreachable!')


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p098_words.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0098.ts
// project euler: problem 98

import { zip } from "@std/collections";
import { assetData } from "../lib/asset.ts";
import { isqrt, maxLst } from "../lib/math.ts";
import { dedupSort, range } from "../lib/util.ts";

class SquareTbl {
  tbl: Map<number, Set<string>>;

  constructor() {
    this.tbl = new Map<number, Set<string>>();
  }

  get(k: number): Set<string> {
    if (this.tbl.has(k)) {
      return this.tbl.get(k)!;
    } else {
      const sqSet = new Set(
        range(isqrt(10 ** (k - 1)), isqrt(10 ** k - 1) + 1)
          .map((x) => x * x)
          .filter((x) => x >= 10 ** (k - 1))
          .map((x) => String(x)),
      );
      this.tbl.set(k, sqSet);

      return sqSet;
    }
  }
}

const findSquares = (w1: string, w2: string, sqSet: Set<string>): number[] => {
  const result: Set<number> = new Set();

  for (const sq of sqSet.values()) {
    for (const n of makeNumber(w1, w2, sq)) {
      if (sqSet.has(n)) {
        result.add(Number(n));
        result.add(Number(sq));
      }
    }
  }

  return Array.from(result.values());
};

const makeNumber = (w1: string, w2: string, sq: string): string[] => {
  const lookup = (ch: string, map: [string, string][]): string => {
    for (const [k, v] of map) {
      if (k === ch) {
        return v;
      }
    }

    return "";
  };

  const transPair = dedupSort(zip(w1.split(""), sq.split("")));
  if (transPair.length !== dedupSort(w1.split("")).length) {
    return [];
  }
  if (transPair.length !== dedupSort(transPair.map(([_, x]) => x)).length) {
    return [];
  }

  return [w2.split("").map((x) => lookup(x, transPair)).join("")];
};

export const compute = (data: string): string => {
  const parseData = (data: string): [number, string, string][] => {
    const keywords = data.replaceAll('"', "").split(",");

    return keywords.map((w) => [w.length, w.split("").sort().join(""), w]);
  };

  const isAnagram = (
    a: [number, string, string],
    b: [number, string, string],
  ): boolean => {
    return a[0] === b[0] && a[1] === b[1];
  };

  let numLst: number[] = [];
  const sqTbl = new SquareTbl();
  const words = parseData(data);
  for (const i of range(0, words.length - 1)) {
    for (const j of range(i + 1, words.length)) {
      if (!isAnagram(words[i], words[j])) {
        continue;
      }
      numLst = numLst.concat(
        findSquares(words[i][2], words[j][2], sqTbl.get(words[i][0])!),
      );
    }
  }

  if (numLst.length === 0) {
    throw new Error("there is no answer");
  } else {
    return String(maxLst(numLst));
  }
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p098_words.txt"));
  return compute(data);
};
